Purchases/Caching/RCDeviceCache.m

```diff
@@ -33,29 +34,58 @@ @interface RCDeviceCache ()
 @implementation RCDeviceCache
 
 - (instancetype)initWith:(NSUserDefaults *)userDefaults {
-    return [self initWith:userDefaults offeringsCachedObject:nil];
+    return [self initWith:userDefaults offeringsCachedObject:nil notificationCenter:nil];
 }
 
 - (instancetype)initWith:(NSUserDefaults *)userDefaults
-   offeringsCachedObject:(RCInMemoryCachedObject<RCOfferings *> *)offeringsCachedObject {
+   offeringsCachedObject:(RCInMemoryCachedObject<RCOfferings *> *)offeringsCachedObject
+      notificationCenter:(NSNotificationCenter *)notificationCenter {
     self = [super init];
     if (self) {
-        if (userDefaults == nil) {
-            userDefaults = [NSUserDefaults standardUserDefaults];
-        }
-        self.userDefaults = userDefaults;
-
         if (offeringsCachedObject == nil) {
             offeringsCachedObject =
                 [[RCInMemoryCachedObject alloc] initWithCacheDurationInSeconds:CACHE_DURATION_IN_SECONDS];
         }
         self.offeringsCachedObject = offeringsCachedObject;
 
+        if (notificationCenter == nil) {
+            notificationCenter = NSNotificationCenter.defaultCenter;
+        }
+        self.notificationCenter = notificationCenter;
+
+        if (userDefaults == nil) {
+            userDefaults = NSUserDefaults.standardUserDefaults;
+        }
+        self.userDefaults = userDefaults;
+        [self observeAppUserIDChanges];
     }
 
     return self;
 }
 
+#pragma mark - UserDefaults Observer
+
+- (void)observeAppUserIDChanges {
+    [self.notificationCenter addObserver:self
+                                selector:@selector(handleUserDefaultsChanged:)
+                                    name:NSUserDefaultsDidChangeNotification
+                                  object:self.userDefaults];
+}
+
+- (void)handleUserDefaultsChanged:(NSNotification *)notification {
+    if (notification.object == self.userDefaults) {
+        if (!self.cachedAppUserID) {
+            NSAssert(false, @"[Purchases] - Cached appUserID has been deleted from user defaults. "
+                            "This leaves the SDK in an undetermined state. Please make sure that RevenueCat "
+                            "entries in user defaults don't get deleted by anything other than the SDK.");
```

Comment:
Do we have any docs or issue in GitHub we could point them too?

Reply:
added, will merge once tests pass